### PR TITLE
Initial implementation of validator disabling strategy

### PR DESCRIPTION
**Pull request by tdimitrov**
The strategy is described in https://github.com/paritytech/polkadot-sdk/issues/784#issuecomment-1691886382. This PR outlines an initial implementation with todos.
Related to https://github.com/paritytech/polkadot-sdk/issues/784
Link to the PR in the old substrate repo: https://github.com/paritytech/substrate/pull/14697

Things that needs to be addressed:

- [ ] Fix the TODOs in the code.
- [ ] Add tests
- [ ] Add runtime migration for staking palet
- [ ] Add benchamarks

### CODE REVIEW

File: substrate/frame/staking/src/pallet/mod.rs [hunk 1]
```diff
@@ -560,19 +566,29 @@ pub mod pallet {
...
+	/// * u32 - The stash account of the offender
+	/// * Perbill - slash proportion
 	#[pallet::storage]
 	#[pallet::unbounded]
 	#[pallet::getter(fn offending_validators)]
-	pub type OffendingValidators<T: Config> = StorageValue<_, Vec<(u32, bool)>, ValueQuery>;
+	pub type OffendingValidators<T: Config> = StorageValue<_, Vec<(u32, Perbill)>, ValueQuery>;
+
+	/// Keep track which validators are disabled because on new session start they should be
+	/// disabled again. The disabled list in `SessionInterface` is cleared on each new session.
```
Comment (tdimitrov) [inline substrate/frame/staking/src/pallet/mod.rs:588]: TODO: https://github.com/paritytech/substrate/pull/14697/files#r1283238804

File: substrate/frame/staking/src/slashing.rs [hunk 2]
```diff
@@ -338,33 +343,136 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
...
+			Err(pos) => {
+				offending.insert(pos, (validator_index_u32, slash_proportion));
 
 				let offending_threshold =
 					T::OffendingValidatorsThreshold::get() * validators.len() as u32;
 
 				if offending.len() >= offending_threshold as usize {
-					// force a new era, to select a new validator set
+					// force a new era, to select a new validator set. The era will be forced on the
+					// next session
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:354]: TODO: https://github.com/paritytech/substrate/pull/14697/files#r1283242773

File: substrate/frame/staking/src/slashing.rs [hunk 3]
```diff
@@ -338,33 +343,136 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
...
-					T::SessionInterface::disable_validator(validator_index_u32);
+					disable_new_offender::<T>(offending, validator_index_u32);
 				}
 			},
-			Ok(index) => {
-				if disable && !offending[index].1 {
+			Ok(pos) => {
+				// Keep the biggest offence
+				if slash_proportion > offending[pos].1 {
+					offending[pos].1 = slash_proportion;
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:365]: TODO: https://github.com/paritytech/substrate/pull/14697/files#r1283244786

File: substrate/frame/staking/src/slashing.rs [hunk 4]
```diff
@@ -338,33 +343,136 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
...
+// Gets a Vec and its desired length as an input. Randomly removes values from the Vec until its
+// length matches the desired length. The validator length should be bigger than the desired one.
+fn remove_random_validators<T: Config>(validators: Vec<u32>, desired_len: usize) -> Vec<u32> {
+	debug_assert!(validators.len() > desired_len);
+
+	let entropy = compute_entropy::<T>(<frame_system::Pallet<T>>::parent_hash());
+	let mut rng = rand_chacha::ChaChaRng::from_seed(entropy.into());
+
+	let mut indecies_to_remove = BTreeSet::new();
+	for _ in 0..=desired_len.saturating_sub(validators.len()) {
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:420]: TODO: https://github.com/paritytech/substrate/pull/14697/files#r1283287217

File: substrate/frame/staking/src/slashing.rs [hunk 5]
```diff
@@ -338,33 +343,136 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
...
+
+	// Now selectively disable based on the offence
+	let mut disabled = Vec::new();
+	for (v, o) in current_offenders {
+		if disable_rand::<T>(o) {
+			disabled.push(*v);
+		}
+	}
+
+	let disabled = if disabled.len() > LIMIT {
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:464]: TODO: https://github.com/paritytech/substrate/pull/14697/files#r1283303368

File: substrate/frame/staking/src/slashing.rs [hunk 6]
```diff
@@ -338,33 +343,136 @@ fn add_offending_validator<T: Config>(stash: &T::AccountId, disable: bool) {
...
+		// remove `disabled.len() - LIMIT` random elements from `disabled`
+		remove_random_validators::<T>(disabled, LIMIT)
+	} else {
+		disabled
+	};
```
Comment (tdimitrov) [inline substrate/frame/staking/src/slashing.rs:469]: TODO: https://github.com/paritytech/substrate/pull/14697/files#r1283297614

File: substrate/frame/staking/src/pallet/mod.rs (hunk 1 above)
Comment (Ank4n) [reply substrate/frame/staking/src/pallet/mod.rs:588]: I believe we will always have an upper bound on `DisabledOffenders`. We should store this in a `BoundedVec<u32, MaxDisabledOffenders>`.

File: substrate/frame/staking/src/slashing.rs (hunk 3 above)
Comment (Ank4n) [inline substrate/frame/staking/src/slashing.rs:365]: Should offence score not add up?

**Comment by tdimitrov** [comment]
bot fmt

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4222349

**Comment by tdimitrov** [comment]
Closing in favour of https://github.com/paritytech/polkadot-sdk/pull/2172

**Pull request closed without merging**